spring-cloud-netflix-eureka-server/src/main/java/org/springframework/cloud/netflix/eureka/server/metrics/EurekaInstanceMetricsBinder.java

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright 2013-2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.netflix.eureka.server.metrics;
+
+import java.util.List;
+import java.util.Objects;
+
+import com.netflix.eureka.registry.PeerAwareInstanceRegistry;
+import io.micrometer.core.instrument.Gauge;
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.Tag;
+import io.micrometer.core.instrument.binder.MeterBinder;
+
+/**
+ * @author heowc
+ */
+public class EurekaInstanceMetricsBinder implements MeterBinder {
+
+	private final PeerAwareInstanceRegistry instanceRegistry;
+
+	EurekaInstanceMetricsBinder(PeerAwareInstanceRegistry instanceRegistry) {
+		this.instanceRegistry = Objects.requireNonNull(instanceRegistry);
+	}
+
+	@Override
+	public void bindTo(MeterRegistry meterRegistry) {
+		instanceRegistry.getApplications().getRegisteredApplications().stream()
+				.flatMap(application -> application.getInstances().stream()).forEach(instanceInfo -> {
+					Gauge.builder("eureka.server.application.instances", () -> 1L)
+							.description("Information about application instances registered on the Eureka server.")
+							.tags(List.of(Tag.of("application", instanceInfo.getAppName()),
+									Tag.of("id", instanceInfo.getId()), Tag.of("host", instanceInfo.getHostName()),
```

Comment:
When it comes to actual metrics, I'll request review from @jonatan-ivanov and you should go along with his suggestions. My preliminary thoughts are that "id" should probably be removed, cause we'll have many singular values, as each instance has a unique id. Also, I'm not sure port is going to be helpful as a tag. While, we might want to slice the data based on how many instances are running on each host, I'm not sure what slicing by port would be useful for.

Reply:
I agree with the "id", if there will be lot of instances, it can be a problem.
I think host+port together can be problematic too because of the same reason: together they are unique per instance (just like the id) so both together have the same problem.

On the other hand, some users might need these so I think it would be useful to inject some kind of a `TagsProvider` that can give you a `Tags` instance for an `instanceInfo` object. There is an issue with this: if we will not provide instance-level tags (dimensions) somehow we need to count them since right now `1` is hardcoded (which works right now since this creates a separate Gauge for every instance).

Also, what everyone thinks about adding `status` as a tag (can be `UP`/`DOWN`)?

Reply:
Right, the status is there already.

Reply:
Ooops, I did not see `status`, sorry. :)
Btw to increase the readability you can also do this:
```java
.description("Information about application instances registered on the Eureka server.")
.tag("application", instanceInfo.getAppName())
.tag("status", instanceInfo.getStatus().name()))
.register(meterRegistry);
```

Reply:
Added `EurekaInstanceTagProvider` 😄